### PR TITLE
Add `minLines` option to text widgets

### LINKED ISSUE #31134: Feature Request: Add minLines to Text (Not TextField)
I've seen many requests from other user to add minLines into **TextField**. But please also add it for **Text**. Thanks

### PR DESCRIPTION
**Pull request by jankuss**
This PR adds the ability to specify `minLines` to text related widgets and classes (`Text`, `RichText`, `Paragraph`, `TextPainter`).
This is useful when we want a text to span a specified number of lines, no matter the actual content of the text. 

Right now, this is possible with some workarounds (see https://github.com/flutter/flutter/issues/31134#issuecomment-560089409), which don't work well when `maxLines` with ellipsis is involved. 

This PR closes #31134

This video shows the implemented feature. All texts have `minLines: 5`.

https://user-images.githubusercontent.com/1659532/167919018-44d1fca2-ea67-42f1-91f1-a643ecef1db1.mov

# Behaviour:
- When no `minLines` is provided, or `minLines` is 0, the feature is disabled
- When `minLines` is provided with a value larger than 0, the text bounds will span the specified amount of `minLines` (using `preferredLineHeight` as the height of a line for reference)
- When `minLines` is used in combination with `maxLines`, the text will stop at the specified `maxLines`. The `minLines` still defines the total height of the text.

# Open questions:

## 1. `minLines` & `maxLines` in combination

Does the behaviour of `minLines` in combination with `maxLines` make sense, when `minLines` is larger than `maxLines`? 

The text will stop at the specified `maxLines`, but the total height of the text will go beyond to have `minLines` of lines.

## 2. Line Height of "filler" lines (when used in `RichText` with mixed font sizes)

Is there a better way to determine the line height for filler lines than `preferredLineHeight` of `TextPainter`? 

The `preferredLineHeight` works great for texts with a uniform font size, but in combination with `RichText` with **mixed font sizes**, this solution doesn't seem optimal. The `preferredLineHeight` for mixed font sizes in the text would change based on the count of lines, which causes the total height of the text to change when a line is added or removed.

In theory, since we are computing the line metrics, we could use the minimum line height or the maximum line height for the text as well - or even make this configurable by the user.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by goderbauer** [comment]
Looks like the analyzer is unhappy with this one, could you please fix it?

**Comment by jankuss** [comment]
Fixed the analyzer issues, now he is happy

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -626,8 +649,43 @@ class TextPainter {
...
+
+  void _layoutMinLines() {
+    if (minLines == null || minLines! == 0) {
+      // If `minLines` is null or equal to zero, no additional height needs to
+      // be added.
+      _additionalMinLinesHeight = 0;
+      return;
+    }
+
+    final List<ui.LineMetrics> lines = computeLineMetrics();
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:664]: Hmm there doesn't seem to be a cheap way to get the total number of lines, or tell `minLines` > # of lines.

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:664]: Ah I don't see any other options.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 2]
```diff
@@ -339,6 +342,25 @@ class TextPainter {
     markNeedsLayout();
   }
 
+
+  /// An optional minimum number of lines for the text to span.
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/painting/text_painter.dart:346]: I don't think we should add `minLines` to `TextPainter`. We already have `TextPainter.preferredLineHeight` so computing the height for `minLines` shouldn't be that difficult without changing the `TextPainter` API. `TextPainter` is pretty much the lowest-level text public API, I'd say we should avoid adding convenience methods like this to keep the interface simple.

Also,  `minLines` doesn't make sense in a rich text context: the first line and the second line can have different line heights.

File: packages/flutter/lib/src/widgets/text.dart [hunk 3]
```diff
@@ -42,13 +42,15 @@ class DefaultTextStyle extends InheritedTheme {
     this.textAlign,
     this.softWrap = true,
     this.overflow = TextOverflow.clip,
+    this.minLines,
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text.dart:45]: Is there a reason we're adding `minLines` to `DefaultTextStyle`?

File: packages/flutter/lib/src/widgets/text.dart [hunk 4]
```diff
@@ -399,6 +413,7 @@ class Text extends StatelessWidget {
     this.softWrap,
     this.overflow,
     this.textScaleFactor,
+    this.minLines,
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text.dart:416]: `minLines` is going to be really strange if the lines have different heights.

File: packages/flutter/lib/src/widgets/text.dart [hunk 5]
```diff
@@ -477,6 +492,12 @@ class Text extends StatelessWidget {
   /// [MediaQuery], or 1.0 if there is no [MediaQuery] in scope.
   final double? textScaleFactor;
 
+  /// An optional minimum number of lines for the text to span.
+  ///
+  /// If the text falls short of the given number of lines, lines will be added
+  /// below until there are the given number of lines.
+  final int? minLines;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text.dart:499]: What's the difference between `minLines == null` and `minLines == 0`?

File: packages/flutter/lib/src/widgets/text.dart [hunk 6]
```diff
@@ -42,13 +42,15 @@ class DefaultTextStyle extends InheritedTheme {
...
     this.overflow = TextOverflow.clip,
+    this.minLines,
     this.maxLines,
     this.textWidthBasis = TextWidthBasis.parent,
     this.textHeightBehavior,
     required super.child,
   }) : assert(style != null),
        assert(softWrap != null),
        assert(overflow != null),
+       assert(minLines == null || minLines >= 0),
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text.dart:53]: Do we allow `minLines > maxLines`?

**Review by LongCatIsLooong — COMMENTED** [review]
Thanks for the pull request and the screen recording!

I think there are workarounds that allow us to deal with the `minLines` problem relatively well without changing the API (e.g., https://github.com/flutter/flutter/issues/31134#issuecomment-1126597446, or create a new `RenderObject` to achieve that).

File: packages/flutter/lib/src/widgets/text.dart (hunk 5 above)
Comment (jankuss) [reply packages/flutter/lib/src/widgets/text.dart:499]: We can make `minLines` default to 0, there is no behavioural difference.

**Comment by jankuss** [comment]
Thanks for the review!

I think the solution you posted in https://github.com/flutter/flutter/issues/31134#issuecomment-1126597446 is the cleanest and doesn't involve tinkering with low level APIs. `minLines` doesn't really make a lot of sense for `RichText`, so only having it for `Text` is the way to go.

I'll close this, since the referenced issue now has a good workaround which solves most use cases.

**Pull request closed without merging**